sound/soc/sof/ipc4-topology.c

```diff
@@ -732,6 +732,13 @@ static int sof_ipc4_widget_setup_comp_mixer(struct snd_sof_widget *swidget)
 	if (ret)
 		goto err;
 
+	ida_init(&swidget->source_queue_ida);
+	ida_init(&swidget->sink_queue_ida);
```

Comment:
Presumably a mux operation will have multiple inputs and one output. 
Likewise a demux will have one input and multiple outputs.

This ida_inits would need to be done in a different place or rely on a helper? This feels too specific to the mixer case, we have other types of widgets that need this.

Reply:
@bardliao I commented on the issue after I posted the patch too. I think adding the ops for get/put queue might not be necessary. The reason I say that is for effect type widgets, some of them might need special handling in terms of queue ID whereas some others might not. So adding the ops based on widget type will be futile. 

So instead what I suggest is adding the array of binding_widget_names per pin for each widget like this:
```
diff --git a/sound/soc/sof/sof-audio.h b/sound/soc/sof/sof-audio.h
index 5ad77b001c36..47de77318557 100644
--- a/sound/soc/sof/sof-audio.h
+++ b/sound/soc/sof/sof-audio.h
@@ -389,6 +389,10 @@ struct snd_sof_widget {
 
        struct snd_sof_module_library_info *lib_info;
 
+       /* array of source/sink widget names that should be bound to each pin */
+       const char *src_queue_binding_widgets[MAX_WIDGET_PINS];
+       const char *sink_queue_binding_widgets[MAX_WIDGET_PINS];
+
        void *private;          /* core does not touch this */
 };
 
```

And then when binding modules, we could just test for the arrays being NULL like below and use the generic queue get/put.
```
diff --git a/sound/soc/sof/ipc4-topology.c b/sound/soc/sof/ipc4-topology.c
index 22ea628d78d0..3ffdcf2cb07e 100644
--- a/sound/soc/sof/ipc4-topology.c
+++ b/sound/soc/sof/ipc4-topology.c
@@ -1500,6 +1500,12 @@ static int sof_ipc4_route_setup(struct snd_sof_dev *sdev, struct snd_sof_route *
        dev_dbg(sdev->dev, "bind %s -> %s\n",
                src_widget->widget->name, sink_widget->widget->name);
 
+       if (!src_widget->src_queue_binding_widgets)
+               sroute->src_queue_id = sof_ipc4_generic_get_queue(swidget);
+
+       if (!sink_widget->sink_queue_binding_widgets)
+               sroute->sink_queue_id = sof_ipc4_generic_get_queue(swidget);
+
        header = src_fw_module->man4_module_entry.id;
        header |= SOF_IPC4_MOD_INSTANCE(src_widget->instance_id);
        header |= SOF_IPC4_MSG_TYPE_SET(SOF_IPC4_MOD_BIND);

```

These araays will be NULL for any widget that does not need any special handling and for those that do, the topology will need to add the binding widget names and the kernel will need parse them.

Reply:
> Presumably a mux operation will have multiple inputs and one output. Likewise a demux will have one input and multiple outputs.

@plbossart I thought about this. But I am not sure can a widget be with multiple input and output?
For example, can a mixer have 2 inputs and 3 outputs?

> 
> This ida_inits would need to be done in a different place or rely on a helper? This feels too specific to the mixer case, we have other types of widgets that need this.

The idea of adding ida_inits in the setup callback is that I set the queue id with ops, and it is tied to widget type. The point is that I don't think all kind of widgets need ida to assign the queue id. For example, PGA should be always with 1 input and 1 output, right? For such widgets, we can always set queue id = 0. @ranj063 This is also the reason why I think using ops is a good idea.

Reply:
For the queue ID we will have several cases:
1. single input and single output, queue == 0 always
2. multiple inputs/outputs, with special queue
3. multiple inputs/output, but no special queue, use incremental queue id.

For 1, I think we can use the num_input_pins/num_output_pins(defined by topology), if num_input_pins/num_output_pins == 1, return 0 as the queue id.
For 2, num_input_pins/num_output_pins has to be greater than 1, and the binding array from topology should be not NULL, we can get the queue number by iterate the binding array.
For 3, num_input_pins/num_output_pins is greater than 1, and binding array is NULL,  we can use ida or even a simple increasing integer `current_queue_index` in swidget

I have no flavor to use helper function or widget ops, but all three cases should be handled together, whatever helper function or widget ops is used, so that the queue support code doesn't distribute everywhere. The function has to have swidget, connected_swidget, route direction in its signature.

My two cents to make the code cohesive.

Reply:
> The idea of adding ida_inits in the setup callback is that I set the queue id with ops, and it is tied to widget type. The point is that I don't think all kind of widgets need ida to assign the queue id. For example, PGA should be always with 1 input and 1 output, right? For such widgets, we can always set queue id = 0. @ranj063 This is also the reason why I think using ops is a good idea.

@bardliao if a widget only has 1 input anf 1 output, ida_alloc() will assign queue ID 0's to it anyway right? Why complicate it with special handling?